src/tangle.ts

```diff
@@ -51,17 +64,59 @@ export class Client<T> {
         return this._inbound.value.transient;
     }
 
+    /**
+     * notify bus of new sandbox
+     */
+    public notify() {
+        this._context.clients.set(this.id, true);
+        this._notifer.next(this._context);
+    }
+
+    /**
+     * collect and hold information about connected sandboxes
+     */
+    private _registerContext() : Observable<Context> {
+        const context$ = this._inbound.pipe(
+            pluck('context'),
+            filter(Boolean),
+            scan((acc, curr) => {
+                if (typeof curr?.clients?.forEach === 'function') {
+                    curr?.clients?.forEach((function(value, key) {
+                        acc.clients.set(key, value);
+                    }));
```

Comment:
> When we talked about this issue my idea was to have the `bus` propagate the latest state to a client once connected fully automatically. This differs from the current behavior as the bus only propagates a state when it changes and not when a client connects. I don't mind going forward with this solution as it solves our issue and maybe exploring mine some time later.

I see. There are two reasons why I'd be hesitant to change the default behavior a) if state/event values are actually time-critical (e.g. a counter starting at 1) it'd be non-deterministic on the client's end if it's a push-on-connect or an emission in real-time and 2) it would fundamentally assume that it's always the right behavior. I think the proposed API changes give consumers the ability to do either. However, I could see how an option to turn on the proposed push last state on connect might be convenient. Implementation would be straightforward using the new API internally. Happy to tackle in as part of a separate issue if we agree on the path forward.

> 
> That said I am a bit confused about the conceptual difference between `registerPromise` and `readyPromise`.

So, `registerPromise` will resolve into `Bus` object once all sandboxes were resolved (async creation in webview's case) & attached whereas `Bus.readyPromise` will fire once every attached sandbox emitted a special `connected` message back to the bus account for time it takes to fully initialize a sandboxes runtime plus construction & attaching the client. I can be more elaborate in the docs if it's still unclear.

> 
> Two suggestions:
> 
> * can we document `registerPromise` and `readyPromise` in the readme?
> * can we highlight the conceptual difference between the two?

Yes and yes. However, maybe we lower confusion with a more descriptive method name as per discussion below?

Reply:
> it'd be non-deterministic

Makes sense. The use case would be here that you have a bus sending data to clients not before they have all connected (as requirement for the bus to even send data in the first place)

> However, I could see how an option to turn on the proposed push last state on connect might be convenient

I was mistakenly thinking that this would be the default behavior already but was wrong with this assumption. I feel like this is the way you'ld expect tangle to work since we put the emphasis on state sharing

>  Implementation would be straightforward using the new API internally. Happy to tackle in as part of a separate issue if we agree on the path forward.

That would be awesome! I am happy to move forward with this change, release it and add the missing piece in a subsequent PR.

Reply:
I think the difficulty here is that depending on what sandboxes one's using the latency varies and, for example, in VSCode, webviews might resolve asynchronously at a later point in time. Specifically when panels are being collapsed/uncollapsed.

I will create an issue for the follow-up work. Let's be sure we get on the same page want this to work before I tackle the implementation.

Reply:
In fact, there's a use-case about handling collapsed/uncollapsed panels with tangle that it currently does not support which i'd like to make part of the discussion. we can pick up the convo in the new ticket.

Reply:
> I will create an issue for the follow-up work. Let's be sure we get on the same page want this to work before I tackle the implementation.

Sounds good to me